openedx_certificates/models.py

```diff
@@ -1,5 +1,345 @@
+        """
+        Get the list of eligible learners for the given course.
+
+        :return: A list of user IDs.
+        """
+        func_path = self.certificate_type.retrieval_func
+        module_path, func_name = func_path.rsplit('.', 1)
+        module = import_module(module_path)
+        func = getattr(module, func_name)
+
+        custom_options = {**self.certificate_type.custom_options, **self.custom_options}
+        return func(self.course_id, custom_options)
+
+    def generate_certificate_for_user(self, user_id: int, celery_task_id: int = 0):
+        """
+        Celery task for processing a single user's certificate.
+
+        This function retrieves an ExternalCertificateCourse object based on course_id and certificate_type_id,
+        retrieves the data using the retrieval_func specified in the associated ExternalCertificateType object,
+        and passes this data to the function specified in the generation_func field.
+
+        Args:
+            user_id: The ID of the user to process the certificate for.
+            celery_task_id (optional): The ID of the Celery task that is running this function.
+        """
+        user = get_user_model().objects.get(id=user_id)
+        # Use the name from the profile if it is not empty. Otherwise, use the first and last name.
+        # We check if the profile exists because it is absent in unit tests.
+        user_full_name = getattr(getattr(user, 'profile', None), 'name', f"{user.first_name} {user.last_name}")
+        custom_options = {**self.certificate_type.custom_options, **self.custom_options}
+
+        try:
+            certificate = ExternalCertificate.objects.get(
+                user_id=user_id,
+                course_id=self.course_id,
+                certificate_type=self.certificate_type.name,
+            )
+            certificate.user_full_name = user_full_name
+            certificate.status = ExternalCertificate.Status.GENERATING
+            certificate.generation_task_id = celery_task_id
+            certificate.save()
+        except ExternalCertificate.DoesNotExist:
+            certificate = ExternalCertificate.objects.create(
+                user_id=user_id,
+                user_full_name=user_full_name,
+                course_id=self.course_id,
+                certificate_type=self.certificate_type.name,
+                status=ExternalCertificate.Status.GENERATING,
+                generation_task_id=celery_task_id,
+            )
+
+        try:
+            generation_module_name, generation_func_name = self.certificate_type.generation_func.rsplit('.', 1)
+            generation_module = import_module(generation_module_name)
+            generation_func = getattr(generation_module, generation_func_name)
+
+            # Run the functions. We do not validate them here, as they are validated in the model's clean() method.
+            certificate.download_url = generation_func(self.course_id, user, certificate.uuid, custom_options)
```

Comment:
Is `download_url` the most generic data to receive back from a generation function? I'm thinking of badges which are more displayed then downloaded. Also external credentials like something on Accredible aren't downloaded either. Maybe just calling it `detail_url` will work though because I assume everything has a URL to view it in detail.

The most general approach would be to let `generation_func` return a dict of JSON data, which in the case of PDFs includes a `download_url` key.

I'm also wondering about how the URLs work for PDF certificates. I assume we want to guarantee the URLs are stable forever, so using the underlying django storage URL of the PDF may not be ideal. Is there some intermediate layer that provides permalinks for each certificate and will redirect or serve the PDF as needed?